frappe/celery_app.py

```diff
@@ -80,17 +82,116 @@ def get_beat_schedule(conf):
 			'schedule': timedelta(seconds=conf.scheduler_interval or DEFAULT_SCHEDULER_INTERVAL)
 		},
 	}
-	
+
 	if conf.celery_queue_per_site:
 		schedule['sync_queues'] = {
 			'task': 'frappe.tasks.sync_queues',
 			'schedule': timedelta(seconds=conf.scheduler_interval or DEFAULT_SCHEDULER_INTERVAL)
 		}
-		
+
 	return schedule
-	
+
 def celery_task(*args, **kwargs):
-	return get_celery().task(*args, **kwargs)
-	
+	func = get_celery().task(*args, **kwargs)
+	return func
+
+def make_async_task(args):
+	task = frappe.new_doc("Async Task")
+	task.update(args)
+	task.status = "Queued"
+	task.set_docstatus_user_and_timestamp()
+	task.db_insert()
+
+def run_test():
+	result = test.delay(site=frappe.local.site)
+	result = result.get(propagate=False)
+	print result
+
+@celery_task()
+def test(site=None):
+	time.sleep(2)
+	raise Exception
+	print "task"
+
+class MonitorThread(object):
+	def __init__(self, celery_app, interval=1):
+		self.celery_app = celery_app
+		self.interval = interval
+
+		self.state = self.celery_app.events.State()
+
+		self.thread = threading.Thread(target=self.run, args=())
+		self.thread.daemon = True
+		self.thread.start()
+
+	def catchall(self, event):
+		if event['type'] != 'worker-heartbeat':
+			self.state.event(event)
+
+			if not 'uuid' in event:
+				return
+
+			task = self.state.tasks.get(event['uuid'])
+			info = task.info()
+
+			if 'enqueue_events_for_site' in event['name']:
+				# don't log enqueue events
+				return
+
+			try:
+				kwargs = eval(info.get('kwargs'))
+
+				if 'site' in kwargs:
+					frappe.connect(kwargs['site'])
+
+					if event['type']=='task-received':
+						make_async_task({'name': event['uuid'], 'task_name': event['name']})
+
+					if event['type']=='task-succeeded':
+						task = frappe.get_doc("Async Task", event['uuid'])
```

Comment:
Let's have a function: `update_async_task(event['uuid'], { "status": "Succeeded", "result": info.get('result') ... })` so that we can update both make & update async task functions if needed without forgetting about the other